File: week10/7/[241104] (Leet375)_Guess_Number_Higher_or_Lower_2.py
"""문제
We are playing the Guessing Game. The game will work as follows:

    1. I pick a number between 1 and n.
    2. You guess a number.
    3. If you guess the right number, you win the game.
    4. If you guess the wrong number, then I will tell you whether the number I picked is higher or lower, and you will continue guessing.
    5. Every time you guess a wrong number x, you will pay x dollars. If you run out of money, you lose the game.
    6. Given a particular n, return the minimum amount of money you need to guarantee a win regardless of what number I pick.

Constraint: 
    1 <= n <= 200
"""
"""Hint : MinMax알고리즘 (turn이 돌아가는 game에서 활용하는 알고리즘)
    -- 각 숫자 선택이 틀렸을 경우, 나머지 범위에 대한 최대비용을 고려하고, 이 최대비용 중 최솟값을 선택하는 것.
    ▶ 이 문제는 MinMax알고리즘과 DP를 결합해서 풀이가능. 
        --MinMax
            1) 최악의 비용 계산 : 숫자 x가 틀렸을 때, 왼쪽범위 [low, x-1]과 오른쪽범위[x+1, high]에서 각각 게임을 계속했을 경우의 최악의 비용 계산. 왼쪽, 오른쪽에서 얻은 값 중 더 큰 값.
            2) 최소화 전략 : 1)에서 계산된 최악의 비용 중 최소가 되는 x를 선택. 즉, 모든 가능한 x 중에서 1)값이 가장 적은 x 찾기.
        --DP
            :DP 사용을 통해 subproblem의 결과를 저장해 중복계산 방지.
"""

"""수도코드1: Recursive
Function getMinimumCostToGuessNumber(n):
    - 정의: n개의 숫자 중에서 최소 비용으로 숫자를 맞추는 비용을 계산
    Create a recursive function called CalculateCost with parameters (start, end):
        - start와 end는 추측해야 할 숫자 범위를 나타냄
        If the range is empty or contains only one number:
            - 범위에 숫자가 없거나 하나만 있는 경우, 비용은 0이 됨
            Return 0
        Initialize minimumCost to a very large number:
            - 최소 비용을 매우 큰 수로 초기화
        For each number x in the range from start to end:
            - 시작부터 끝까지 범위 내의 각 숫자 x에 대해
            Calculate the cost if x is the wrong guess:
                - x가 잘못된 추측일 경우의 비용 계산
                - 왼쪽 범위(start to x-1)와 오른쪽 범위(x+1 to end) 중 더 큰 비용을 계산
            Add the cost of guessing x to the above cost:
                - 추측한 x의 비용을 위에서 계산한 비용에 더함            
            Update minimumCost if the current cost is lower:
                - 현재 비용이 더 낮을 경우, 최소 비용을 업데이트
        Return the minimumCost:
            - 계산된 최소 비용을 반환

    Call CalculateCost(1, n) to get the minimum cost for the range 1 to n:
        - 범위 1부터 n에 대한 최소 비용을 얻기 위해 CalculateCost를 호출
    Return the result of CalculateCost:
        - CalculateCost의 결과를 반환
"""

#1) Recursive (minmax) : Time limit exceeded
def getMoneyAmount(n):
    def dfs(low, high):
        if low >= high : 
            return 0
        min_cost = float('inf')
        for x in range(low, high):
            cost = x + max(dfs(low, x-1), dfs(x+1, high))
            min_cost = min(cost, min_cost)
        return min_cost
    return dfs(1,n)


"""수도코드2 : Top down DP
★ 범위 기반의 DP는 2차원 배열을 사용하는 것이 일반적임.각 범위의 시작점, 끝점.
Function CalculateMinimumCost(n):
    - 입력 n은 게임에서 사용될 최대 숫자를 나타냄
    Initialize a cache to store results of subproblems:
        - 하위 문제의 결과를 저장하기 위한 캐시(메모리) 초기화
    Define a recursive function called CostToGuess(start, end):
        - 숫자 범위를 입력으로 받아 최소 비용을 계산하는 재귀 함수 정의
        If start is greater than or equal to end:
            - 범위에 숫자가 하나도 없거나 하나만 있는 경우
            Return 0
            - 비용은 0
        If the result for (start, end) is in the cache:
            - 현재 범위에 대한 결과가 캐시에 이미 저장되어 있는지 확인
            Return the cached result
            - 저장된 결과를 반환
        Initialize minimumCost to infinity:
            - 최소 비용을 무한대로 초기화
        For each possible guess x from start to end:
            - 가능한 모든 추측 x에 대해 반복
            Calculate the maximum cost if x is not correct:
                - x가 틀렸을 경우에 대비하여 최대 비용 계산
                leftCost = Call CostToGuess(start, x-1)
                - 왼쪽 범위에 대한 비용
                rightCost = Call CostToGuess(x+1, end)
                - 오른쪽 범위에 대한 비용
            totalCost = x + max(leftCost, rightCost)
                - 추측 비용과 더 큰 비용을 합산
            Update minimumCost if totalCost is less:
                - totalCost가 현재의 minimumCost보다 작으면 업데이트
        Store the minimumCost in the cache for (start, end):
            - 계산된 최소 비용을 캐시에 저장
        Return minimumCost
            - 최소 비용 반환

    Call CostToGuess(1, n) to get the minimum cost from 1 to n:
        - 1부터 n까지의 최소 비용을 계산하기 위해 CostToGuess 호출
    Return the result from the recursive function
        - 재귀 함수로부터의 결과를 반환
"""

# 2) Top Down
def getMoneyAmount(n):
    n_mincost_result = [[-1 for _ in range(n+1)] for _ in range(n+1)]  #2차원으로 만들어야함!
    
    def dfs(low, high, n_result):
        if low >= high :
            return 0
        if n_result[low][high] != -1 :
            return n_result[low][high]
        min_cost = float('inf')
        for x in range(low, high):
            maximumcost = x + max(dfs(low, x-1, n_result), dfs(x+1, high, n_result))
            min_cost = min(min_cost, maximumcost)
        n_result[low][high] = min_cost
        return min_cost
    
    return  dfs(1, n, n_mincost_result)


"""수도코드3: Bottom-up DP
Function CalculateMinimumCost(n):
    - n: 게임에서 사용할 최대 숫자

    Initialize a 2D array dp where dp[i][j] will hold the minimum cost to guess from number i to j:
        - i에서 j까지 추측하는 데 필요한 최소 비용을 저장할 2차원 배열 dp를 초기화

    For each length from 2 to n (inclusive):
        - 길이 2부터 n까지 각 길이에 대해
        For each start from 1 to n-length+1:
            - 시작점 1부터 n-length+1까지
            Set end to start+length-1:
                - 끝점을 start+length-1로 설정
            Initialize dp[start][end] to a large number:
                - dp[start][end]를 큰 숫자로 초기화
            For each guess k from start to end:
                - start부터 end까지 각 추측 k에 대해
                Calculate the cost if guess k is wrong:
                    - 추측 k가 틀렸을 경우의 비용을 계산
                    cost = k + max(dp[start][k-1], dp[k+1][end] if k+1 <= end else 0)
                Update dp[start][end] with the minimum of its current value and cost:
                    - dp[start][end]를 현재 값과 비용 중 작은 값으로 업데이트

    Return dp[1][n]:
        - 1부터 n까지 추측하는 데 필요한 최소 비용 반환
"""

# 3) Bottom-up
def getMoneyAmount(n):
    mincostDP = [ [0 for _ in range(n+1)] for _ in range(n+1)]
    
    for length in range(2, n+1):
        for start in range(1, n-length+1 +1):
            end = start + length - 1
            mincostDP[start][end] = float('inf')
            for guess_num in range(start, end+1):
                # 오른쪽 범위가 유효한 경우와 그렇지 않은 경우를 분리하여 처리
                if guess_num == end:
                    cost = guess_num + mincostDP[start][guess_num-1]
                else:
                    cost = guess_num + max(mincostDP[start][guess_num-1], mincostDP[guess_num+1][end])
                    mincostDP[start][end] = min(cost, mincostDP[start][end])
    
    return mincostDP[1][n]
    

File: week10/7/[241104] (Leet375)_Guess_Number_Higher_or_Lower_2.cpp
/*문제
We are playing the Guessing Game. The game will work as follows:

    1. I pick a number between 1 and n.
    2. You guess a number.
    3. If you guess the right number, you win the game.
    4. If you guess the wrong number, then I will tell you whether the number I picked is higher or lower, and you will continue guessing.
    5. Every time you guess a wrong number x, you will pay x dollars. If you run out of money, you lose the game.
    6. Given a particular n, return the minimum amount of money you need to guarantee a win regardless of what number I pick.

Constraint: 
    1 <= n <= 200
*/

/*수도코드1 : Recursive
    1. X 숫자가 들어왔을 때의 이기기 위해 필요한 최대 돈 = x +  MAX((LOW, x-1), (x+1, HIGH))
    2. x를 돌면서, "각 X에서의 필요한 최대 돈" 중 MIN 값 찾아 반환.
*/
#include <iostream>
using namespace std;
#define MAX_INT 2147483647

class Solution{
public:
    int getMoneyAmountHelp(int low, int high){
        if (low >= high) return 0;

        int min_cost = MAX_INT;

        for (int x = low ; x <= high ; x++){
            int low_to_x = getMoneyAmountHelp(low, x-1);
            int x_to_high = getMoneyAmountHelp(x+1, high);
            int max_cost = low_to_x > x_to_high ? low_to_x : x_to_high;
            int cost = x + max_cost;
            if (min_cost > cost) min_cost = cost;
        }
        return min_cost;

    }
    int getMoneyAmount(int n){
        return getMoneyAmountHelp(1, n);
    }
};


/*-------------------------------------------------------------------------------------------------------------------------------------------*/
/*수도코드2 : Top down
    1. low, high 각각 축으로 가지는 2D vector인 DPT정의
    2. 2D array의 값을 -1로 초기화
    3. 만약 DPT[low][high] 값이 -1이면 recursive로 직접 구하고
        DPT[low][high] 값이 -1이 아니면 저장된 값 사용.
*/

#include <iostream>
#include <vector>
using namespace std;
#define MAX_INT 2147483647

class Solution{
public:
    int getMoneyAmountHelp(int low, int high, vector<vector<int>> &DPT){
        if (low >= high) return 0;

        if (DPT[low][high] != -1) return DPT[low][high];

        int min_cost = MAX_INT;
        for (int x = low ; x <= high ; x++){
            int low_to_x = getMoneyAmountHelp(low, x-1, DPT);
            int x_to_high = getMoneyAmountHelp(x+1, high, DPT);
            int max_cost = low_to_x > x_to_high ? low_to_x : x_to_high;
            int cost = x + max_cost;
            if (min_cost > cost) min_cost = cost;
        }
        DPT[low][high] = min_cost;
        return DPT[low][high];

    }
    int getMoneyAmount(int n){
        vector<vector<int>> DPT = {}; 
        //vector<vector<int>> DPT(n+1, vector<int>(n+1, -1));  //이렇게 한줄로 초기선언하는 것도 가능.
        for (int i = 0; i < n+1; i++) {
            vector<int> DPT_row(n+1, -1);
            DPT.push_back(DPT_row);
        }

        return getMoneyAmountHelp(1, n, DPT);
    }
};

/*-------------------------------------------------------------------------------------------------------------------------------------------*/
/*수도코드3 : Bottom up
    1. low, high 각각 축으로 가지는 2D vector인 DPT정의
    2. low, high 간의 범위를 2~10까지 조정해가면서 
        해당 범위 내에서의 start, end에 따른 min_cost 값을 DPT에 저장.
    3. return DPT[1][n]
*/
#include <iostream>
#include <vector>
using namespace std;
#define MAX_INT 2147483647

class Solution{
public:
    int getMoneyAmount(int n){
        vector<vector<int>> DPT(n+1, vector<int>(n+1, 0));  //이렇게 한줄로 초기선언하는 것도 가능.

        for (int length = 2; length <= n; length++){
            for (int start = 1; start <= n-length+1; start++){
                int end = start + length -1;
                DPT[start][end] = MAX_INT;
                for (int x = start; x <= end; x++) {
                    int cost;
                    if (x == end) { // x가 범위의 끝이면 오른쪽에 더이상 고려할 숫자 없음. (x==start일 경우는 어차피 자동으로 이 됨.)
                        cost = x + DPT[start][x-1];  // x가 end와 같을 때는 오른쪽 범위가 없으므로
                    } else {
                        int left = DPT[start][x-1];  // 왼쪽 범위의 최대 비용
                        int right = DPT[x+1][end];  // 오른쪽 범위의 최대 비용
                        cost = x + max(left, right);  // x를 선택했을 때의 총 비용
                    }
                    DPT[start][end] = min(DPT[start][end], cost);  // 최소 비용 갱신
                }
            }
        }
        return DPT[1][n];  // 1부터 n까지의 최소 비용 반환
    }
};


/* 추가 설명 참고
코드에서 x == end 조건을 특별히 검사하는 이유는 선택된 숫자 x가 범위의 끝일 때 오른쪽에 더 이상 고려할 숫자가 없기 때문입니다. 이 경우, 추가적인 비용은 왼쪽 범위의 최대 비용과 x 자신의 비용만을 고려하면 됩니다. 오른쪽 범위는 존재하지 않으므로, 그 비용은 0입니다.

그러나 x == start인 경우를 별도로 처리하지 않는 이유는 다음과 같습니다:

비대칭적 조건: 문제의 특성 상, x를 선택할 때의 비용 계산은 x를 기준으로 왼쪽 범위(start부터 x-1)와 오른쪽 범위(x+1부터 end)의 최대 비용을 고려합니다. x가 start와 같을 때는 왼쪽 범위가 없기 때문에, 왼쪽 범위의 최대 비용은 자동으로 0입니다. 이는 코드 로직에서 자연스럽게 처리됩니다. 즉, x-1이 start-1이 되므로, 이 범위는 유효하지 않으며, 이 때의 최대 비용은 0으로 간주됩니다.

코드 단순화: x == start를 명시적으로 처리하는 것은 필요하지 않습니다. 왜냐하면 이 경우의 왼쪽 비용은 자동으로 0이 되기 때문입니다. 이를 명시적으로 코드에 쓰지 않아도 DPT[start][x-1]는 유효 범위를 벗어나는 인덱스 접근이 아닙니다. Python에서는 리스트 접근이 범위를 벗어나면 오류를 발생시키지만, C++에서는 벡터의 특정 인덱스 접근이 범위를 벗어나면 초기화되지 않은 메모리를 읽을 수 있으므로, 초기화된 동적 프로그래밍 테이블을 사용하여 이러한 문제를 방지합니다.

결국, 이 코드는 x의 위치가 start이든 end이든 각 위치에 따른 적절한 조건 처리를 통해 문제 요구사항을 효과적으로 충족합니다. x == start에서는 왼쪽 비용이 자동으로 0이 되고, x == end에서는 오른쪽 비용이 0이 됩니다. 이러한 방식으로 모든 가능한 위치에서 최소 비용을 찾아내는 로직을 구현합니다.*/